File: evd_ros_backend/evd_ros_core/src/evd_script/test/container.py

from ..node import Node


class Container(Node):

    '''
    Data structure methods
    '''

    def __init__(self, item_type, values=[], type='', name='', uuid=None, parent=None, append_type=True):

        self._values = None
        self._item_type = None

        super(Container,self).__init__(
            type='container<{}>.'.format(itemType) + type if append_type else type,
            name=name,
            uuid=uuid,
            parent=parent,
            append_type=append_type)

        self.values = values
        self.item_type = item_type

    def to_dct(self):
        msg = super(Container,self).to_dct()
        msg.update({
            'values': [x.to_dct() for x in self.values],
            'item_type': self.item_type
        })
        return msg

    @classmethod
    def from_dct(cls, dct):
        from ..utility_functions import NodeParser

        return cls(
            item_type=dct['item_type'],
            values=[NodeParser(x) for x in dct['values']],
            type=dct['type'],
            append_type=False,
            name=dct['name'],
            uuid=dct['uuid']
        )

    '''
    Data accessor/modifier methods
    '''

    @property
    def values(self):
        return self._values

    @values.setter
    def values(self, value):
        if self._values != value:

            for v in self._values:
                v.remove_from_cache()

            self._values = values
            for v in self._values:
                v.parent = self

            self.updated_attribute('values','set')

    @property
    def item_type(self):
        return _item_type

    @item_type.setter
    def item_type(self, value):
        if self._item_type != value:
            self._item_type = value
            self.updated_attribute('item_type','set')

    def add(self, value):
        value.parent = self
        self._values.append(value)
        self.updated_attribute('values','add',value.uuid)

    def get(self, uuid):
        return self._values[self.search_for_index(uuid)]

    def delete(self, uuid):
        idx = self.search_for_index(uuid)
        if idx == -1:
            raise Exception('Value not in container')

        self._values[idx].remove_from_cache()
        self._values.pop(idx)

        self.updated_attribute('values','delete',uuid)

    def search_for_index(self, uuid):
        idx = -1
        for i in range(0,len(self.values)):
            if self.values[i].uuid == uuid:
                idx = i
                break
        return idx

    def set(self, dct):
        from ..utility_functions import NodeParser

        if 'item_type' in dct.keys():
            self.item_type = dct.get('item_type')

        if 'values' in dct.keys():
            self.values = [NodeParser(x) for x in dct['values']]

        super(Container,self).set(dct)

    '''
    Cache Methods
    '''

    def remove_from_cache(self):
        for v in self._values:
            v.remove_from_cache()

        super(Container,self).remove_from_cache()

    def add_to_cache(self):
        for v in self._values:
            v.add_to_cache()

        super(Container,self).add_to_cache()

    '''
    Children Methods
    '''

    def delete_child(self, uuid):
        success  = False

        if uuid in [v.uuid for v in self._values]:
            self.delete(uuid)
            success = True

        return success

    '''
    Update Methods
    '''

    def late_construct_update(self):

        for v in self.values:
            v.late_construct_update()

        super(Container,self).late_construct_update()

    def deep_update(self):

        for v in self.values:
            v.deep_update()

        super(Container,self).deep_update()

        self.updated_attribute('values', 'update')
        self.updated_attribute('item_type', 'update')

    def shallow_update(self):
        super(Container,self).shallow_update()

        self.updated_attribute('values', 'update')
        self.updated_attribute('item_type', 'update')

File: evd_unity_sim_app/Assets/Application/_Scripts/EvD Script Model/test/Container.cs
using System.Collections;
using System.Collections.Generic;


namespace EvD
{
    namespace Test 
    {

        [System.Serializable]
        public class Container<T> : Node where T : Node
        {

            /*
            * Private Members
            */

            private string _itemType = null;

            private List<T> _values = null;

            /*
            * Constructors
            */

            public static new string TypeString()
            {
                return string.Format("container<{0}>.", typeof(T).ToString());
            }

            public static new string FullTypeString()
            {
                return Node.FullTypeString() + TypeString();
            }

            public Container(List<T> list = null, string itemType = null, string type = "", string name = "",
                            string uuid = null, Node parent = null, bool appendType = true)
            : base(appendType ? string.Format("container<{0}>.", typeof(T).ToString()) + type : type, name, uuid, parent, appendType)
            {
                values = (list != null) ? list : new List<T>();
                itemType = (itemType != null) ? itemType : typeof(T).ToString();
            }

            public static new Container<T> FromDict(Dictionary<string,object> dct)
            {
                var vals = new List<T>();
                foreach (var v in (List<Dictionary<string,object>>)dct["values"])
                {
                    vals.Add((T)typeof(T).GetMethod("FromDict").Invoke(null, new object[] { v }));
                }

                return new Container<T>(
                    list: vals,
                    itemType: (string)dct["item_type"],
                    type: (string)dct["type"],
                    name: (string)dct["name"],
                    uuid: (string)dct["uuid"],
                    appendType: false
                );
            }

            public override Dictionary<string, object> ToDict()
            {
                var valsDct = new List<Dictionary<string, object>>();
                foreach (var v in values)
                {
                    valsDct.Add(v.ToDict());
                }

                var dct = base.ToDict();
                dct.Add("values", valsDct);
                dct.Add("item_type", itemType);
                return dct;
            }

            /*
            * Accessors and Modifiers
            */

            public string itemType 
            {
                get
                {
                    return _itemType;
                }

                set
                {
                    if (_itemType != value) {
                        _itemType = value;
                        UpdatedAttribute("item_type","set");
                    }
                }
            }

            public List<T> values
            {
                get
                {
                    return _values;
                }

                set
                {
                    if (_values != value)
                    {
                        if (_values != null)
                        {
                            foreach (var v in _values)
                            {
                                v.RemoveFromCache();
                            }
                        }

                        _values = value;
                        foreach (var v in _values)
                        {
                            v.parent = this;
                        }

                        UpdatedAttribute("values", "set");
                    }
                }
            }

            public void Add(T v)
            {
                v.parent = this;
                values.Add(v);
                UpdatedAttribute("values", "add", v.uuid);
            }

            public T Get(string uuid)
            {
                return values[SearchForIndex(uuid)];
            }

            public void Delete(string uuid)
            {
                int idx = SearchForIndex(uuid);
                if (idx == -1)
                {
                    throw new System.Exception("Value not in container");
                }

                values[idx].RemoveFromCache();
                values.RemoveAt(idx);

                UpdatedAttribute("values", "delete", uuid);
            }

            public int SearchForIndex(string uuid)
            {
                int idx = -1;
                for (int i=0; i<values.Count; i++)
                {
                    if (values[i].uuid == uuid)
                    {
                        idx = i;
                        break;
                    }
                }
                return idx;
            }

            public override void Set(Dictionary<string, object> dct)
            {
                if (dct.ContainsKey("values"))
                {
                    var vals = new List<T>();
                    foreach (var v in (List<Dictionary<string,object>>)dct["values"])
                    {
                        vals.Add((T)typeof(T).GetMethod("FromDict").Invoke(null, new object[] { v }));
                    }
                    values = vals;
                }

                if (dct.ContainsValue("item_type")) {
                    itemType = (string)dct["item_type"];
                }

                base.Set(dct);
            }

            /*
            * Cache Methods
            */

            public override void RemoveFromCache()
            {
                foreach (var v in values)
                {
                    v.RemoveFromCache();
                }
                base.RemoveFromCache();
            }

            public override void AddToCache()
            {
                foreach (var v in values)
                {
                    v.AddToCache();
                }
                base.AddToCache();
            }

            /*
            * Children Methods
            */

            public override bool DeleteChild(string uuid)
            {
                bool success = false;

                foreach (var v in values)
                {
                    if (v.uuid == uuid)
                    {
                        Delete(uuid);
                        success = true;
                        break;
                    }
                }

                return success;
            }

            /*
            * Update Methods
            */

            public override void LateConstructUpdate()
            {
                foreach (var v in values)
                {
                    v.LateConstructUpdate();
                }

                base.LateConstructUpdate();
            }

            public override void DeepUpdate()
            {
                foreach (var v in values)
                {
                    v.DeepUpdate();
                }

                base.DeepUpdate();

                UpdatedAttribute("values", "update");
            }

            public override void ShallowUpdate()
            {
                base.ShallowUpdate();

                UpdatedAttribute("values", "update"); 
            }
        }

    }
}
